File: lightroom_tagger/core/error_policy.py
"""Pluggable error escalation policies for FallbackDispatcher."""

from __future__ import annotations

from enum import Enum
from typing import Any, Protocol, runtime_checkable

from lightroom_tagger.core.exceptions import (
    ContextLengthError,
    InvalidRequestError,
    PayloadTooLargeError,
    RateLimitError,
)

MAX_TOKENS_ESCALATION = [256, 4096, 32768, 65536]
BATCH_MAX_TOKENS_ESCALATION = [4096, 32768, 65536]

RATE_LIMIT_ABORT_THRESHOLD = 3
FATAL_ABORT_THRESHOLD = 3


class ConsecutiveAbortTracker:
    """Session-level consecutive rate-limit / fatal counters for vision scoring.

    Owned by the dispatcher and consulted before each dispatch.  The scoring
    loop reads ``fatal_abort_reached`` to stop orchestrating candidates; it
    does not re-inspect ``RateLimitError`` / ``InvalidRequestError``.
    """

    def __init__(
        self,
        *,
        rate_limit_threshold: int = RATE_LIMIT_ABORT_THRESHOLD,
        fatal_threshold: int = FATAL_ABORT_THRESHOLD,
    ) -> None:
        self._rate_limit_threshold = rate_limit_threshold
        self._fatal_threshold = fatal_threshold
        self._consecutive_rate_limits = 0
        self._consecutive_fatal = 0

    @property
    def consecutive_rate_limits(self) -> int:
        return self._consecutive_rate_limits

    @property
    def consecutive_fatal(self) -> int:
        return self._consecutive_fatal

    @property
    def rate_limit_abort_reached(self) -> bool:
        return self._consecutive_rate_limits >= self._rate_limit_threshold

    @property
    def fatal_abort_reached(self) -> bool:
        return self._consecutive_fatal >= self._fatal_threshold

    def record_success(self) -> None:
        self._consecutive_rate_limits = 0

    def record_rate_limit(self) -> None:
        self._consecutive_rate_limits += 1
        self._consecutive_fatal = 0

    def record_fatal(self) -> None:
        self._consecutive_rate_limits = 0
        self._consecutive_fatal += 1

    def record_transient_error(self) -> None:
        self._consecutive_rate_limits = 0
        self._consecutive_fatal = 0

    def record_dispatch_outcome(self, exc: Exception | None) -> None:
        """Update counters after a dispatch completes (success or final failure)."""
        if exc is None:
            self.record_success()
        elif isinstance(exc, RateLimitError):
            self.record_rate_limit()
        elif isinstance(exc, InvalidRequestError):
            self.record_fatal()
        else:
            self.record_transient_error()


class EscalationAction(Enum):
    """What to do after an escalation-class error."""

    RETRY = "retry"
    GIVE_UP = "give_up"
    SPLIT = "split"


@runtime_checkable
class ErrorPolicy(Protocol):
    """Hook invoked on retryable escalation-class errors."""

    def on_escalation_error(
        self,
        exc: Exception,
        *,
        provider_id: str,
        model: str,
        operation: str,
        call_state: dict[str, Any],
    ) -> EscalationAction:
        """Decide retry-with-mutation vs give-up.

        Policies may mutate ``call_state`` (and their own instance state)
        before returning ``EscalationAction.RETRY``.
        """


class NoOpErrorPolicy:
    """Default policy — no mutation; normal retry / fallback behaviour."""

    def on_escalation_error(
        self,
        exc: Exception,
        *,
        provider_id: str,
        model: str,
        operation: str,
        call_state: dict[str, Any],
    ) -> EscalationAction:
        return EscalationAction.RETRY


class ContextLengthEscalationPolicy:
    """Vision compare max_tokens ladder with per-run cache and blacklist."""

    def __init__(self, ladder: list[int] | None = None) -> None:
        self._ladder = list(ladder if ladder is not None else MAX_TOKENS_ESCALATION)
        self._model_min_tokens: dict[str, int] = {}
        self._broken_provider_models: set[str] = set()

    @property
    def ladder(self) -> list[int]:
        return list(self._ladder)

    @property
    def model_min_tokens(self) -> dict[str, int]:
        return dict(self._model_min_tokens)

    @property
    def broken_provider_models(self) -> set[str]:
        return set(self._broken_provider_models)

    def provider_key(self, provider_id: str, model: str) -> str:
        return f"{provider_id}:{model}"

    def is_broken(self, provider_id: str, model: str) -> bool:
        return self.provider_key(provider_id, model) in self._broken_provider_models

    def starting_index(self, provider_id: str, model: str) -> int:
        key = self.provider_key(provider_id, model)
        cached_min = self._model_min_tokens.get(key, 0)
        for i, val in enumerate(self._ladder):
            if val >= cached_min:
                return i
        return 0

    def max_tokens_at(self, index: int) -> int:
        return self._ladder[index]

    def on_escalation_error(
        self,
        exc: Exception,
        *,
        provider_id: str,
        model: str,
        operation: str,
        call_state: dict[str, Any],
    ) -> EscalationAction:
        if not isinstance(exc, ContextLengthError):
            return EscalationAction.RETRY

        key = self.provider_key(provider_id, model)
        idx = int(call_state.get("token_index", 0))

        if idx < len(self._ladder) - 1:
            new_idx = idx + 1
            next_val = self._ladder[new_idx]
            self._model_min_tokens[key] = next_val
            call_state["token_index"] = new_idx
            call_state["_log_message"] = (
                f"[{operation}] Escalating max_tokens to {next_val} for {model}"
            )
            return EscalationAction.RETRY

        self._broken_provider_models.add(key)
        call_state["_log_message"] = (
            f"[{operation}] max_tokens exhausted at {self._ladder[idx]} "
            f"for {model}, blacklisting for session"
        )
        return EscalationAction.GIVE_UP


class VisionBatchErrorPolicy:
    """Batch vision: context-length ladder plus payload split on 413."""

    def __init__(
        self,
        token_ladder: list[int] | None = None,
        token_policy: ContextLengthEscalationPolicy | None = None,
    ) -> None:
        ladder = token_ladder if token_ladder is not None else BATCH_MAX_TOKENS_ESCALATION
        self._token_policy = (
            token_policy
            if token_policy is not None
            else ContextLengthEscalationPolicy(ladder=ladder)
        )

    @property
    def ladder(self) -> list[int]:
        return self._token_policy.ladder

    @property
    def model_min_tokens(self) -> dict[str, int]:
        return self._token_policy.model_min_tokens

    @property
    def broken_provider_models(self) -> set[str]:
        return self._token_policy.broken_provider_models

    def provider_key(self, provider_id: str, model: str) -> str:
        return self._token_policy.provider_key(provider_id, model)

    def is_broken(self, provider_id: str, model: str) -> bool:
        return self._token_policy.is_broken(provider_id, model)

    def starting_index(self, provider_id: str, model: str) -> int:
        return self._token_policy.starting_index(provider_id, model)

    def max_tokens_at(self, index: int) -> int:
        return self._token_policy.max_tokens_at(index)

    def on_escalation_error(
        self,
        exc: Exception,
        *,
        provider_id: str,
        model: str,
        operation: str,
        call_state: dict[str, Any],
    ) -> EscalationAction:
        if isinstance(exc, PayloadTooLargeError):
            candidates = call_state.get("candidates") or []
            if len(candidates) <= 1:
                call_state["_log_message"] = (
                    f"[{operation}] single-item chunk still too large, skipping"
                )
                return EscalationAction.GIVE_UP

            half = len(candidates) // 2
            call_state["_split_halves"] = (candidates[:half], candidates[half:])
            call_state["_log_message"] = (
                f"[{operation}] payload too large, splitting "
                f"{len(candidates)} -> {half}+{len(candidates) - half}"
            )
            return EscalationAction.SPLIT

        return self._token_policy.on_escalation_error(
            exc,
            provider_id=provider_id,
            model=model,
            operation=operation,
            call_state=call_state,
        )

File: apps/visualizer/backend-ts/src/providers/error-policy.ts
/**
 * Pluggable error escalation policies for the fallback dispatcher.
 * Port of `core/error_policy.py`.
 *
 * Two separate concerns live here. `ConsecutiveAbortTracker` is a *session*
 * counter — it exists so a batch of 40,000 images stops after three consecutive
 * rate limits instead of grinding through the whole queue collecting 429s. The
 * policies are *per-call*: they mutate the call state and ask for a retry, which
 * is how the max_tokens ladder and the 413 payload split work.
 */
import { ContextLengthError, InvalidRequestError, PayloadTooLargeError, RateLimitError } from './errors.js';

export const MAX_TOKENS_ESCALATION = [256, 4096, 32768, 65536];
export const BATCH_MAX_TOKENS_ESCALATION = [4096, 32768, 65536];

export const RATE_LIMIT_ABORT_THRESHOLD = 3;
export const FATAL_ABORT_THRESHOLD = 3;

/**
 * Session-level consecutive rate-limit and fatal counters.
 *
 * Owned by the dispatcher and consulted before each dispatch. The scoring loop
 * reads `fatalAbortReached` to stop orchestrating candidates rather than
 * re-inspecting the error types itself.
 */
export class ConsecutiveAbortTracker {
  private readonly rateLimitThreshold: number;
  private readonly fatalThreshold: number;
  private rateLimits = 0;
  private fatal = 0;

  constructor(
    opts: { rateLimitThreshold?: number; fatalThreshold?: number } = {},
  ) {
    this.rateLimitThreshold = opts.rateLimitThreshold ?? RATE_LIMIT_ABORT_THRESHOLD;
    this.fatalThreshold = opts.fatalThreshold ?? FATAL_ABORT_THRESHOLD;
  }

  get consecutiveRateLimits(): number {
    return this.rateLimits;
  }

  get consecutiveFatal(): number {
    return this.fatal;
  }

  get rateLimitAbortReached(): boolean {
    return this.rateLimits >= this.rateLimitThreshold;
  }

  get fatalAbortReached(): boolean {
    return this.fatal >= this.fatalThreshold;
  }

  /** A success clears the rate-limit streak but deliberately not the fatal one. */
  recordSuccess(): void {
    this.rateLimits = 0;
  }

  recordRateLimit(): void {
    this.rateLimits += 1;
    this.fatal = 0;
  }

  recordFatal(): void {
    this.rateLimits = 0;
    this.fatal += 1;
  }

  recordTransientError(): void {
    this.rateLimits = 0;
    this.fatal = 0;
  }

  /** Update the counters once a dispatch has settled, successfully or not. */
  recordDispatchOutcome(e: unknown): void {
    if (e === null || e === undefined) this.recordSuccess();
    else if (e instanceof RateLimitError) this.recordRateLimit();
    else if (e instanceof InvalidRequestError) this.recordFatal();
    else this.recordTransientError();
  }
}

/** What to do after an escalation-class error. */
export type EscalationAction = 'retry' | 'give_up' | 'split';

/** Mutable per-call state a policy may modify before asking for a retry. */
export type CallState = Record<string, unknown>;

export interface ErrorPolicy {
  onEscalationError(
    e: unknown,
    ctx: { providerId: string; model: string; operation: string; callState: CallState },
  ): EscalationAction;
}

/** No mutation — the plain retry and fallback behaviour. */
export class NoOpErrorPolicy implements ErrorPolicy {
  onEscalationError(): EscalationAction {
    return 'retry';
  }
}

/**
 * Climb a `max_tokens` ladder on a context-length error, then blacklist.
 *
 * The per-session cache matters on a batch: once a model is known to need 4096
 * tokens, every later image starts there instead of failing at 256 first. And a
 * model that fails at the top of the ladder is blacklisted for the session, so
 * the batch stops paying for a call that cannot succeed.
 */
export class ContextLengthEscalationPolicy implements ErrorPolicy {
  private readonly ladderValues: number[];
  private readonly modelMinTokens = new Map<string, number>();
  private readonly brokenProviderModels = new Set<string>();

  constructor(ladder?: readonly number[]) {
    this.ladderValues = [...(ladder ?? MAX_TOKENS_ESCALATION)];
  }

  get ladder(): number[] {
    return [...this.ladderValues];
  }

  get minTokens(): Map<string, number> {
    return new Map(this.modelMinTokens);
  }

  get broken(): Set<string> {
    return new Set(this.brokenProviderModels);
  }

  providerKey(providerId: string, model: string): string {
    return `${providerId}:${model}`;
  }

  isBroken(providerId: string, model: string): boolean {
    return this.brokenProviderModels.has(this.providerKey(providerId, model));
  }

  /** The ladder index to start at, given what this model has needed before. */
  startingIndex(providerId: string, model: string): number {
    const cachedMin = this.modelMinTokens.get(this.providerKey(providerId, model)) ?? 0;
    const index = this.ladderValues.findIndex((val) => val >= cachedMin);
    return index === -1 ? 0 : index;
  }

  maxTokensAt(index: number): number {
    return this.ladderValues[index]!;
  }

  onEscalationError(
    e: unknown,
    ctx: { providerId: string; model: string; operation: string; callState: CallState },
  ): EscalationAction {
    if (!(e instanceof ContextLengthError)) return 'retry';

    const key = this.providerKey(ctx.providerId, ctx.model);
    const idx = Math.trunc(Number(ctx.callState.token_index ?? 0));

    if (idx < this.ladderValues.length - 1) {
      const newIdx = idx + 1;
      const nextVal = this.ladderValues[newIdx]!;
      this.modelMinTokens.set(key, nextVal);
      ctx.callState.token_index = newIdx;
      ctx.callState._log_message =
        `[${ctx.operation}] Escalating max_tokens to ${nextVal} for ${ctx.model}`;
      return 'retry';
    }

    this.brokenProviderModels.add(key);
    ctx.callState._log_message =
      `[${ctx.operation}] max_tokens exhausted at ${this.ladderValues[idx]} ` +
      `for ${ctx.model}, blacklisting for session`;
    return 'give_up';
  }
}

/** Batch vision: the context-length ladder plus a payload split on 413. */
export class VisionBatchErrorPolicy implements ErrorPolicy {
  private readonly tokenPolicy: ContextLengthEscalationPolicy;

  constructor(opts: { tokenLadder?: readonly number[]; tokenPolicy?: ContextLengthEscalationPolicy } = {}) {
    this.tokenPolicy =
      opts.tokenPolicy ??
      new ContextLengthEscalationPolicy(opts.tokenLadder ?? BATCH_MAX_TOKENS_ESCALATION);
  }

  get ladder(): number[] {
    return this.tokenPolicy.ladder;
  }

  get minTokens(): Map<string, number> {
    return this.tokenPolicy.minTokens;
  }

  get broken(): Set<string> {
    return this.tokenPolicy.broken;
  }

  providerKey(providerId: string, model: string): string {
    return this.tokenPolicy.providerKey(providerId, model);
  }

  isBroken(providerId: string, model: string): boolean {
    return this.tokenPolicy.isBroken(providerId, model);
  }

  startingIndex(providerId: string, model: string): number {
    return this.tokenPolicy.startingIndex(providerId, model);
  }

  maxTokensAt(index: number): number {
    return this.tokenPolicy.maxTokensAt(index);
  }

  onEscalationError(
    e: unknown,
    ctx: { providerId: string; model: string; operation: string; callState: CallState },
  ): EscalationAction {
    if (e instanceof PayloadTooLargeError) {
      const candidates = (ctx.callState.candidates as unknown[] | undefined) ?? [];
      // A single item that is still too large cannot be split further.
      if (candidates.length <= 1) {
        ctx.callState._log_message =
          `[${ctx.operation}] single-item chunk still too large, skipping`;
        return 'give_up';
      }
      const half = Math.floor(candidates.length / 2);
      ctx.callState._split_halves = [candidates.slice(0, half), candidates.slice(half)];
      ctx.callState._log_message =
        `[${ctx.operation}] payload too large, splitting ` +
        `${candidates.length} -> ${half}+${candidates.length - half}`;
      return 'split';
    }
    return this.tokenPolicy.onEscalationError(e, ctx);
  }
}
